Unstoppabledomains resolution

JohnnyJumper opened the pull request:
Handles #

Changes proposed in this pull request:
- Added ability to use unstoppable domains in adding safe option.
- Similar to ENS

Main network call comes from [@unstoppabledomains/resolution-java](https://github.com/unstoppabledomains/resolution-java) library. Simple https call to blockchain providers such as infura. Can be configured separately at the Resolution builder step.

By default using [linkpool](https://linkpool.io/) service. 

@gnosis/mobile-devs

tschubotz commented:
@JohnnyJumper Thanks for the contribution!

We discussed it briefly yesterday, internally. There were some open questions about UX and the used libs that we wanted to clarify first, internally.

@posthnikova in this PR it's suggested to add "Unstoppable domain" to this action sheet that we have to enter the Safe address: https://zpl.io/aMJDlkz. All screens are essentially similar to the ENS ones (e.g. https://zpl.io/2GJKJ8Y)

Is this okay for you or do you prefer something else from a UX perspective? An alternative would be to use the same input field for both ENS and unstoppable domains (.crypto), however probably more implementation effort.

posthnikova commented:
@tschubotz I'm fine with adding this. There has to be an icon for Android, something like this.

![image](https://user-images.githubusercontent.com/45787536/105193894-4f803d80-5b39-11eb-8036-1afc20976604.png)

tschubotz commented:
@elgatovital @biafra23 in case there still time in the current sprint, could you please review the PR and comment if this matches you expectations etc?
If not, I think we should just plan this regularly for the next sprint afterwards 🙂

https://github.com/gnosis/safe-react/pull/1764#issue-552727175 has test domains for mainnet and rinkeby.

DeRain commented:
> @elgatovital @biafra23 in case there still time in the current sprint, could you please review the PR and comment if this matches you expectations etc?
> If not, I think we should just plan this regularly for the next sprint afterwards 🙂
> 
> [gnosis/safe-react#1764 (comment)](https://github.com/gnosis/safe-react/pull/1764#issue-552727175) has test domains for mainnet and rinkeby.

Hi, do you have updates with the PR? Do you need additional changes from Unstoppable? Thanks!

biafra23 commented:
@DeRain The dependency `com.unstoppabledomains:resolution:1.9.1` pulls `lombok-maven-plugin`. Is that really necessary at runtime?

JohnnyJumper commented:
@biafra23  Thank you for catching this.

We have just released a new version 1.10.0 where we moved Lombok dependency into the compileOnly group. This should remove it from the runtime dependency list. 

Give maven some time to update itself and see if it fixes the issue.

biafra23 commented:
@JohnnyJumper The latest version doesn't compile. There is a parenthesis missing in Line 16 of UnstoppableDomainsRepository.kt

Also a dependency seems to be missing. so `com.unstoppabledomains` is not found in the same file.

biafra23 commented:
@JohnnyJumper @DeRain 
While testing it would be nice to have a safe on Rinkeby, that is registered with UD. Do you have an example name for me?

DeRain commented:
> @JohnnyJumper @DeRain
> While testing it would be nice to have a safe on Rinkeby, that is registered with UD. Do you have an example name for me?

Fee free to get your name using [developers page](https://unstoppabledomains.com/developers) - works for Mainnet and Rinkedby.

DeRain commented:
> @JohnnyJumper @DeRain
> While testing it would be nice to have a safe on Rinkeby, that is registered with UD. Do you have an example name for me?

Or just tell me safe address and i will create and setup domain for you.

biafra23 commented:
Please link this safe on Rinkeby `0x1C8b9B78e3085866521FE206fa4c1a67F49f153A` to **creek**(.crypto(?)) if possible.

DeRain commented:
> Please link this safe on Rinkeby `0x1C8b9B78e3085866521FE206fa4c1a67F49f153A` to **creek**(.crypto(?)) if possible.

It's done. Domain name: udtestdev-creek.crypto

biafra23 commented:
@DeRain udtestdev-creek.crypto doesn't seem to be registered. Do I need to use a specific infiura key?  

Do you have a more interactive channel to resolve this? Could you join our Discord server (https://discord.gg/4ZX5c8gk) and find me (Dirk #4150) there?

Error | Endpoints
-----| ----------
![image](https://user-images.githubusercontent.com/246473/108034052-3766e580-7035-11eb-91c7-398e09e61f0d.png) | ![image](https://user-images.githubusercontent.com/246473/108034080-451c6b00-7035-11eb-8df9-9de4d75bc914.png)

JohnnyJumper commented on data/src/main/java/io/gnosis/data/repositories/UnstoppableDomainsRepository.kt:
```diff
@@ -0,0 +1,36 @@
+    var address = ""
+    var domain: String;
+    private val library: Resolution = Resolution.builder()
+            .infura(NamingServiceType.CNS, asString(getKey("INFURA_API_KEY", ""))
```
I am not really sure how to get this key from the system.
It might be the reason why `udtestdev-creek.crypto` doesn't resolve. 
I have tried to use the BuildConfig, however, it seems the INFURA_API_KEY is not added for this package.

biafra23 replied:
I hard coded the key for now. But you are right the code above doesn't compile/work. So any Infura key should work?

biafra23 replied:
ATM `BuildConfig.INFURA_API_KEY` is not available in the `data` module. Maybe we should change that.

JohnnyJumper replied:
I think i know why the domain is not resolving. can you try to resolve brad.crypto? 
I believe the library is connected to the mainnet by default. You can overwrite this line
`.chainId(NamingServiceType.CNS, Network.ROPSTEN)`

I don't think we have any rinkeby contract deployed, but will ask tomorrow morning with the team

biafra23 replied:
To get the INFURA_API_KEY from `$basedir/project_keys` Please add the following in `data/build.gradle` in the `defaulConfig` section. Right below `NATIVE_CURRENCY_NAME`:
```
        buildConfigField javaTypes.STRING, "INFURA_API_KEY", asString(getKey("INFURA_API_KEY", ""))
```

Then you can use `BuilConfig.INFURA_API_KEY` whereever you need it.

JohnnyJumper replied:
@biafra23  Can you test it out? I checked and was very wrong that we don't have rinkeby contracts. In my mind, I was thinking more of a ropsten than rinkeby. 

The commit 80cee73 resolves rinkeby udtestdev-creek.crypto 
The commit 222f84c resolves mainnet domain such as brad.crypto

DeRain replied:
@biafra23 Are you detecting network id somewhere (mainnet or ropsten)? If yes - how can we access it in the code?

@JohnnyJumper The other way if to detect network on library initialisation by using net_version RPC call: https://eth.wiki/json-rpc/API#net_version

biafra23 replied:
> The commit 80cee73 resolves rinkeby udtestdev-creek.crypto
Adding the chainId works. 

We do not distinguish between network ids ATM. We just use a different service URL. Currently we support `mainnet`, `rinkeby` and `xdai`. Do you support xdai? 

You could look for `rinkeby` in `BuildConfig.CLIENT_GATEWAY_URL` and use mainnet in any other case.

biafra23 replied:
Since they are all different flavors we could also have an explicit `chainId`.

DeRain replied:
@JohnnyJumper

`BuildConfig.BLOCKCHAIN_NET_URL` contains the Infura URL. Please add this to the data module and use it as provider. Please, detect network id on library init (need to update library version).

cc: @biafra23

DeRain commented:
> @DeRain udtestdev-creek.crypto doesn't seem to be registered. Do I need to use a specific Infura key?
> 
> Do you have a more interactive channel to resolve this? Could you join our Discord server (https://discord.gg/4ZX5c8gk) and find me (Dirk #4150) there?
> 
> Error	Endpoints
> ![image](https://user-images.githubusercontent.com/246473/108034052-3766e580-7035-11eb-91c7-398e09e61f0d.png)	![image](https://user-images.githubusercontent.com/246473/108034080-451c6b00-7035-11eb-8df9-9de4d75bc914.png)

Sorry, i'm getting an error when trying to accept invite. Can you try to join our discord - https://discord.com/invite/b6ZVxSZ9Hn ?

I'm `DeRain#7309` here or `Kirill | beresnev.crypto`

JohnnyJumper commented:
@biafra23  I have updated the resolution library to v1.12.0, so that now we can know which network is being used from the provider itself by making a net_version call to the provider.
I also used `BuildConfig.BLOCKCHAIN_NET_URL` with the `BuildConfig.INFURA_API_KEY` as we need a full-provider URL to make the calls.  Would be so kind to test it out?
It did resolve udtestdev-creek.crypto for me with default settings.

biafra23 commented:
Regarding dependencies:
It seems `commons-math3` is not used in your project.
Is it correct, that `jackson` is only used to parse the network config?

biafra23 commented on app/build.gradle:
```diff
@@ -297,6 +297,9 @@ dependencies {
     profileImplementation "com.squareup.leakcanary:leakcanary-android:$versions.leak_canary"
+
+    // Unstoppabledomains
+    implementation 'com.unstoppabledomains:resolution:1.10.0'
```
Should this be version 1.12.0 also?

biafra23 commented on app/src/main/res/layout/dialog_unstoppable_input.xml:
```diff
@@ -0,0 +1,159 @@
+        style="@style/Toolbar"
+        android:gravity="center_vertical"
+        android:orientation="horizontal"
+        app:layout_constraintBottom_toTopOf="@id/dialog_ens_input_url_layout"
```
This should reference: `dialog_unstoppable_domain_layout` instead

biafra23 commented on app/src/main/res/layout/dialog_unstoppable_input.xml:
```diff
@@ -0,0 +1,159 @@
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:fitsSystemWindows="true"
+    android:background="@color/background">
```
This should be `surface_01` instead of `background`

biafra23 commented on app/src/main/java/io/gnosis/safe/ui/dialogs/UnstoppableInputViewModel.kt:
```diff
@@ -0,0 +1,30 @@
+import io.gnosis.data.repositories.UnstoppableDomainsRepository
+import pm.gnosis.model.Solidity
+import javax.inject.Inject
+import com.unstoppabledomains.exceptions.*;
```
Several of these imports are not necessary.

biafra23 commented on app/src/main/java/io/gnosis/safe/ui/dialogs/UnstoppableInputDialog.kt:
```diff
@@ -0,0 +1,149 @@
+import android.view.ViewGroup
+import androidx.core.widget.doOnTextChanged
+import androidx.lifecycle.lifecycleScope
+import com.unstoppabledomains.exceptions.ns.NSExceptionCode
```
This import seems unnecessary.

biafra23 commented on app/src/main/res/layout/bottom_sheet_address_input.xml:
```diff
@@ -117,4 +116,38 @@
+        android:paddingLeft="8dp"
+        android:paddingRight="8dp"
+        android:text="@string/address_input_unstoppabledomains"
+        android:textColor="@color/text_emphasis_medium"
```
This needs to be removed.

biafra23 commented on app/src/main/res/layout/bottom_sheet_address_input.xml:
```diff
@@ -117,4 +116,38 @@
+
+    <TextView
+        android:id="@+id/bottom_sheet_address_input_unstoppabledomains"
+        style="@style/Text"
```
This should be `TextDarkRegular`

biafra23 commented on app/src/main/res/values/strings.xml:
```diff
@@ -342,6 +364,7 @@
     <string name="tx_list_contract_interaction">Contract interaction</string>
     <string name="tx_list_conflict_header_explainer">These transactions conflict as they use the same nonce. Executing one will automatically replace the other(s).</string>
     <string name="tx_list_conflict_header_learn_more">Learn more</string>
+    <string name="address_input_unstoppabledomains">Unstoppable domain</string>
```
For consistency I'd rename this to `address_input_unstoppable_domain`.

JohnnyJumper commented:
@biafra23  we have updated the library and removed the common-math dep as it was not used anymore. Thank you so much for noticing it. We also have added a linter to keep our dependency in check. 

I have updated all the UI aspects as well. Please let me know if there is anything else needed for this.

biafra23 commented:
How can I trigger the CLA? Maybe via this comment?

elgatovital commented:
please update this branch. Rebase it on top of our current gnosis:master or merge master here. Afterwards we need you to accept the CLA.

elgatovital commented on data/src/main/java/io/gnosis/data/repositories/UnstoppableDomainsRepository.kt:
```diff
@@ -0,0 +1,38 @@
+import java.util.concurrent.Callable
+import java.util.concurrent.FutureTask
+
+class BackgroundTask : Callable<String> {
```
this should not be needed, because repository's resolve function is marked as suspend. So please delete this BackgroundTask

elgatovital commented on data/src/main/java/io/gnosis/data/repositories/UnstoppableDomainsRepository.kt:
```diff
@@ -0,0 +1,38 @@
+    }
+}
+
+class UnstoppableDomainsRepository() {
```
please inject DomainResolution as a dependency to the repository. Provide DomainResolution in ApplicationModule

JohnnyJumper replied:
@elgatovital There has been a minor issue with injecting the library as a dependency to the repository. 

The library has been making a net_version call to the provider to get the correct network for self-configure which contract address to use. 
I couldn't found any way to make an async injection with dagger2, if anything I found this thread on [stackoverflow](https://stackoverflow.com/questions/57008801/how-combine-dagger2-and-coroutines) which is saying it is not possible to use suspend with the dagger, as it has no builtin suspend functionality. I am not very familiar with the dagger framework and not sure if there is any workaround for this.

Because of this, I parsed manually BuildConfig.BLOCKCHAIN_NET_URL to get the network from the first word of the URL assuming it is going to be infura. I found this very hacky and it probably won't work with other blockchain providers in case if you will switch to something else. 

Is there any better solution for getting the network?

elgatovital commented on data/src/main/java/io/gnosis/data/repositories/UnstoppableDomainsRepository.kt:
```diff
@@ -0,0 +1,38 @@
+
+class UnstoppableDomainsRepository() {
+    suspend fun resolve(domain: String): Solidity.Address {
+        val futureTask: FutureTask<String> = FutureTask<String>(BackgroundTask(domain))
```
please remove this future task. This function is suspending, so it is possible to wait for the resolution using async / await.
```kotlin
val address = async {  library.getAddress(domain, "eth") }.await()
```

JohnnyJumper replied:
I have done it and Android Studio suggested using Kotlin's `withContext(Dispatchers.IO) {}` instead of async / await.
I believe it acts similarly to async { ... }.await()

elgatovital commented on app/build.gradle:
```diff
@@ -297,6 +297,9 @@ dependencies {
     profileImplementation "com.squareup.leakcanary:leakcanary-android:$versions.leak_canary"
+
+    // Unstoppabledomains
+    implementation 'com.unstoppabledomains:resolution:1.12.1'
```
please add version to versions.gradle file and convert this to 
```
implementation 'com.unstoppabledomains:resolution:$versions.unstoppabledomains'
```

JohnnyJumper replied:
done

DeRain commented:
@biafra23 @elgatovital Hey, thank you for review!
Just to clarify - do you expect mentioned updates from Unstoppable?

elgatovital commented:
> Just to clarify - do you expect mentioned updates from Unstoppable?

@DeRain yes

JohnnyJumper commented:
I have read the CLA Document and I hereby sign the CLA

JohnnyJumper commented:
I have read the CLA Document and I hereby sign the CLA

elgatovital reviewed (commented):
please commit small adjustment to build.gradle to trigger our build server pipeline, which was previously disabled for external branches

JohnnyJumper commented:
Hi @elgatovital, I have updated the correct build.gradle and it did trigger the build pipeline in github CI. It is showing as failing and I am not sure why as I was able to launch and test the integration without any problems locally on my device. 

I don't have any access to the details page to check it out. Whenever I click on it, I got 
`Sorry, the page you’re trying to access could not be found` message. 

Is there something else I should do?

elgatovital commented:
@JohnnyJumper pipeline is failing because we have a threshold for our test coverage and you haven't covered UnstuppableDomainsRepository. Please add unit tests

JohnnyJumper commented:
@elgatovital  should be good now

elgatovital commented:
Hi @JohnnyJumper, base branch has to be updated. You could give me write access to this fork and I will do the rest myself

JohnnyJumper commented:
@elgatovital, I have sent the invitation
